### PR TITLE
chore(e2e): extend logging test to 3 node cluster

### PR DESCRIPTION
**Pull request by ghouscht**
See also https://github.com/etcd-io/etcd/issues/17329

This is how I would extend the logging test to include both a single node and a three node cluster. If it is ok like that I'd submit PRs for backporting as written in the linked issue

Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

### CODE REVIEW

File: tests/e2e/logging_test.go [hunk 1]
```diff
@@ -26,39 +26,60 @@ import (
...
+				e2e.WithLogLevel("debug"),
+			)
+			require.NoError(t, err)
+			defer epc.Close()
+
+			require.Lenf(t, epc.Procs, tc.clusterSize, "embedded etcd cluster process count is not as expected")
+
+			var lines []string
+			for i := range tc.clusterSize {
+				lines = append(lines, epc.Procs[i].Logs().Lines()...)
```
Comment (serathius) [inline tests/e2e/logging_test.go:62]: Think we need to collect log lines after we close the server.

File: tests/e2e/logging_test.go [hunk 2]
```diff
@@ -26,39 +26,60 @@ import (
...
-	ctx := context.TODO()
+	tests := []struct {
+		name          string
+		clusterSize   int
+		allowedErrors map[string]bool
+	}{
+		{
+			name:          "single node cluster",
+			clusterSize:   1,
+			allowedErrors: map[string]bool{"setting up serving from embedded etcd failed.": true},
```
Comment (serathius) [inline tests/e2e/logging_test.go:37]: Would be good to followup why we get this log.

Comment (serathius) [reply tests/e2e/logging_test.go:37]: cc @jmhbnz @ahrtr

File: tests/e2e/logging_test.go (hunk 1 above)
Comment (ghouscht) [reply tests/e2e/logging_test.go:62]: I did encounter the allowed error but apparently not all the time. It indeed does make a difference if `.Lines()` is called after closing. By doing so the resulting log lines seem to be constant. Will update the PR.

File: tests/e2e/logging_test.go (hunk 2 above)
Comment (ghouscht) [reply tests/e2e/logging_test.go:37]: Yes, saw the comment in the issue. From what I saw in `release-3.5` branch this does not happen. I'll have a look where exactly it comes from.

Comment (serathius) [reply tests/e2e/logging_test.go:37]: Interesting, thanks @ghouscht !

Comment (ahrtr) [reply tests/e2e/logging_test.go:37]: The error logs are generated when closing/shutting down etcd. The listeners are closed firstly, then the accepter will return errors. They should be expected. One possible additional test case is to verify that there is no any error logs before closing etcd, which of course can be discussed & implemented separately.

```
{"level":"info","ts":"2024-10-30T08:25:35.363864Z","caller":"osutil/interrupt_unix.go:65","msg":"received signal; shutting down","signal":"interrupt"}
{"level":"info","ts":"2024-10-30T08:25:35.363960Z","caller":"embed/etcd.go:398","msg":"closing etcd server","name":"default","data-dir":"default.etcd","advertise-peer-urls":["http://localhost:2380"],"advertise-client-urls":["http://localhost:2379"]}
......
{"level":"error","ts":"2024-10-30T08:25:35.364157Z","caller":"embed/etcd.go:874","msg":"setting up serving from embedded etcd failed.","error":"http: Server closed","stacktrace":"go.etcd.io/etcd/server/v3/embed.(*Etcd).errHandler\n\tgo.etcd.io/etcd/server/v3/embed/etcd.go:874\ngo.etcd.io/etcd/server/v3/embed.(*serveCtx).serve.func3\n\tgo.etcd.io/etcd/server/v3/embed/serve.go:176"}
{"level":"error","ts":"2024-10-30T08:25:35.364250Z","caller":"embed/etcd.go:874","msg":"setting up serving from embedded etcd failed.","error":"accept tcp 127.0.0.1:2379: use of closed network connection","stacktrace":"go.etcd.io/etcd/server/v3/embed.(*Etcd).errHandler\n\tgo.etcd.io/etcd/server/v3/embed/etcd.go:874\ngo.etcd.io/etcd/server/v3/embed.(*Etcd).serveClients.func1\n\tgo.etcd.io/etcd/server/v3/embed/etcd.go:776"}
...
{"level":"error","ts":"2024-10-30T08:25:35.394822Z","caller":"embed/etcd.go:874","msg":"setting up serving from embedded etcd failed.","error":"accept tcp 127.0.0.1:2380: use of closed network connection","stacktrace":"go.etcd.io/etcd/server/v3/embed.(*Etcd).errHandler\n\tgo.etcd.io/etcd/server/v3/embed/etcd.go:874\ngo.etcd.io/etcd/server/v3/embed.(*Etcd).servePeers.func3\n\tgo.etcd.io/etcd/server/v3/embed/etcd.go:614"}
```

Comment (ghouscht) [reply tests/e2e/logging_test.go:37]: Yeah this is what I figured out as well. In that case these errors are expected and we could simply adapt the [`errorHandler`](https://github.com/ghouscht/etcd/blob/issue-17329/server/embed/etcd.go#L873-L875) function to ignore them, e.g. like this:
```go
	if err != nil &&
		!errors.Is(err, http.ErrServerClosed) &&
		!errors.Is(err, net.ErrClosed) {
		e.GetLogger().Error("setting up serving from embedded etcd failed.", zap.Error(err))
	}
```

Comment (ahrtr) [reply tests/e2e/logging_test.go:37]: It's difficult to be certain whether `http.ErrServerClose` or `net.ErrClosed` is expected by just one error itself. Suggest not to change the production code for now.

**Review by ahrtr — APPROVED** [review]
LGTM

Thanks

File: tests/e2e/logging_test.go (hunk 2 above)
Comment (serathius) [reply tests/e2e/logging_test.go:37]: But I think it is. https://cs.opensource.google/go/go/+/refs/tags/go1.23.2:src/net/http/server.go;l=3286-3288;drc=773767def0e0f29584a69bd760430167b7479d7d 
http.ErrServerClose means.

**Pull request merged**